url29ff2e53f3_nicolas3470_Settlers_of_Catan_AI_tgz-pJ8-robotGameJ8/src/soc/robot/SOCRobotDM.java

```diff
@@ -0,0 +1,2114 @@
+	  threatenedRoad.resetScore();
+	  // float wgetaScore = getWinGameETABonusForRoad(threatenedRoad, currentBuildingETAs[SOCBuildingSpeedEstimate.ROAD], leadersCurrentWGETA, playerTrackers);
+
+	  D.ebugPrintln("$$$  final score = "+threatenedRoad.getScore());
+					
+	  if (favoriteRoad.getCoordinates() != threatenedRoad.getCoordinates()) {
+	    if (secondFavoriteRoad == null) {
+	      secondFavoriteRoad = threatenedRoad;
+	    } else {
+	      if (threatenedRoad.getScore() > secondFavoriteRoad.getScore()) {
+	      secondFavoriteRoad = threatenedRoad;
+	      }
+	    }
+	  }
+	}
+	goodRoadEnum = goodRoads.elements();
+	while (goodRoadEnum.hasMoreElements()) {
+	  SOCPossibleRoad goodRoad = (SOCPossibleRoad)goodRoadEnum.nextElement();
+	  D.ebugPrintln("$$$ good road at "+Integer.toHexString(goodRoad.getCoordinates()));
+	  //
+	  // see how building this piece impacts our winETA
+	  //
+	  goodRoad.resetScore();
+	  // float wgetaScore = getWinGameETABonusForRoad(goodRoad, currentBuildingETAs[SOCBuildingSpeedEstimate.ROAD], leadersCurrentWGETA, playerTrackers);
+
+	  D.ebugPrintln("$$$  final score = "+goodRoad.getScore());
+
+	  if (favoriteRoad.getCoordinates() != goodRoad.getCoordinates()) {
+	    if (secondFavoriteRoad == null) {
+	      secondFavoriteRoad = goodRoad;
+	    } else {
+	      if (goodRoad.getScore() > secondFavoriteRoad.getScore()) {
+		secondFavoriteRoad = goodRoad;
+	      }
+	    }
+	  }
+	}
+				
+	SOCPlayerTracker.undoTryPutPiece(tmpRoad, game);
+
+	if (!buildingPlan.empty()) {
+	  SOCPossiblePiece planPeek = (SOCPossiblePiece)buildingPlan.peek();
+	  if ((planPeek == null) ||
+	      (planPeek.getType() != SOCPlayingPiece.ROAD)) {
+	    if (secondFavoriteRoad != null) {
+	      D.ebugPrintln("### SECOND FAVORITE ROAD IS AT "+Integer.toHexString(secondFavoriteRoad.getCoordinates()));
+	      D.ebugPrintln("###   WITH A SCORE OF "+secondFavoriteRoad.getScore());
+	      D.ebugPrintln("$ PUSHING "+secondFavoriteRoad);
+	      buildingPlan.push(secondFavoriteRoad);
+	      D.ebugPrintln("$ PUSHING "+favoriteRoad);
+	      buildingPlan.push(favoriteRoad);
+	    }
+	  } else if (secondFavoriteRoad != null) {
+	    SOCPossiblePiece tmp = (SOCPossiblePiece)buildingPlan.pop();
+	    D.ebugPrintln("$ POPPED OFF");
+	    D.ebugPrintln("### SECOND FAVORITE ROAD IS AT "+Integer.toHexString(secondFavoriteRoad.getCoordinates()));
+	    D.ebugPrintln("###   WITH A SCORE OF "+secondFavoriteRoad.getScore());
+	    D.ebugPrintln("$ PUSHING "+secondFavoriteRoad);
+	    buildingPlan.push(secondFavoriteRoad);
+	    D.ebugPrintln("$ PUSHING "+tmp);
+	    buildingPlan.push(tmp);
+	  }
+	}     
+      } 
+    } 
+    //long endTime = System.currentTimeMillis();
+    //System.out.println("plan time: "+(endTime-startTime));
+  }
+  
+  /**
+   * dumbFastGameStrategy
+   * uses rules to determine what to build next
+   *
+   * @param buildingETAs  the etas for building something
+   */
+  protected void dumbFastGameStrategy(int[] buildingETAs)
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 75. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.